code/datums/helper_datums/butchering.dm

```diff
@@ -349,3 +352,249 @@ var/global/list/animal_butchering_products = list(
+	var/toolName = null
+	var/speed_mod = 0
+	var/list/butchValues = butcherValueStep(user)
+	if(!butchValues?.len)
+		return
+	if(butchValues[BUTCHER_SPEED])
+		speed_mod = butchValues[BUTCHER_SPEED]
+	if(butchValues[BUTCHER_TOOL])
+		butcherTool = butchValues[BUTCHER_TOOL]
+	if(butchValues[BUTCHER_TNAME])
+		toolName = butchValues[BUTCHER_TNAME]
+	var/list/butcherOptions = butcherMenuStep(user)
+	if(!butcherOptions?.len)
+		return
+	var/datum/butchering_product/typeOfCarve = butcherChooseStep(user, butcherOptions, butcherTool)
+	if(!typeOfCarve)
+		return
+	if(butcherCarveStep(user, speed_mod, typeOfCarve, toolName))
+		if(typeOfCarve != BUTCHER_MEAT)
+			butcherProduct(user, typeOfCarve)
+		else
+			butcherMeat(user, toolName)
+
+
+/mob/living/proc/butcherCheck(mob/user, var/ourTool = null)
+	if(!istype(user))
+		return FALSE
+	if(user.isUnconscious() || user.restrained())
+		return FALSE
+	if(!Adjacent(user))
+		return FALSE
+	if(being_butchered)
+		to_chat(user, "<span class='notice'>[src] is already being butchered.</span>")
+		return FALSE
+	if(!can_butcher)
+		to_chat(user, "<span class='notice'>You can't butcher [src]!")
+		return FALSE
+	if(ourTool && ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if(ourTool != H.get_active_hand())
+			return FALSE
+	return TRUE
+
+
+/mob/living/proc/butcherValueStep(mob/user)
+	var/butchSpeed = 0
+	var/obj/item/theTool = null
+	var/toolName = null
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		theTool = H.get_active_hand()
+		if(theTool)
+			butchSpeed = theTool.is_sharp()
+			toolName = theTool.name
+			if(!butchSpeed)
+				to_chat(user, "<span class='notice'>You can't butcher \the [src] with this!</span>")
+				return
+		if(H.organ_has_mutation(LIMB_HEAD, M_BEAK))
+			var/obj/item/mask = H.get_item_by_slot(slot_wear_mask)
+			if(!mask || !(mask.body_parts_covered & MOUTH)) //If our mask doesn't cover mouth, we can use our beak to help us while butchering
+				butchSpeed += 0.25
+				if(!toolName)
+					toolName = "beak"
+		if(H.organ_has_mutation(H.get_active_hand_organ(), M_CLAWS))
+			if(!istype(H.gloves))
+				butchSpeed += 0.25
+				if(!toolName)
+					toolName = "claws"
+		if(isgrue(H))
+			toolName = "grue"
+			butchSpeed += 0.5
+	else
+		butchSpeed = 0.5
+	if(!butchSpeed)
+		return
+	return list(BUTCHER_SPEED = butchSpeed, BUTCHER_TOOL = theTool, BUTCHER_TNAME = toolName)
+
+
+/mob/living/proc/butcherMenuStep(mob/user)
+	var/list/butcherType = list()
+	if(meat_type && meat_amount > meat_taken)
+		butcherType += "Butcher"
+	for(var/datum/butchering_product/BP in butchering_drops)
+		if(BP.amount)
+			butcherType += BP.verb_name
+	if(!butcherType.len)
+		to_chat(user, "<span class='notice'>There's nothing to butcher.</span>")
+		return
+	butcherType += "Cancel"
+	return butcherType
+
+
+/mob/living/proc/butcherChooseStep(mob/user, butcherOptions, butcherTool)
+	var/choice = input(user,"What would you like to do with \the [src]?","Butchering") in butcherOptions
```

Comment:
```suggestion
	var/choice = input(user,"What would you like to do with \the [src]?","Butchering") in null|butcherOptions
```
then check for null

Reply:
What does that do? Proofing from butcherOptions being null I assume?

Reply:
Gives you a cancel button

Reply:
Oh, ya okay that's pretty important.